### ISSUE TITLE
ONNX inference with dropout

### ISSUE THREAD

**Issue by weixsong** [onnx/onnx#3753]
I have a model based on self-attention, I'd like to enable dropout in ONNX model inference.

But I found that when I set <code>training=torch.onnx.TrainingMode.TRAINING,</code>, and inference with onnxruntime,  the model output are identical for each inference with the same input.

here is the code I used to export model:
```python
    torch.onnx.export(model,  # model being run
                      onnx_input,  # model input (or a tuple for multiple inputs)
                      onnx_path,  # where to save the model (can be a file or file-like object)
                      training=torch.onnx.TrainingMode.TRAINING,  # export in Training mode
                      verbose=True,
                      export_params=True,  # store the trained parameter weights inside the model file
                      opset_version=12,  # the ONNX version to export the model to
                      do_constant_folding=False,  # whether to execute constant folding for optimization
                      input_names=input_names,  # the model's input names
                      output_names=output_names,  # the model's output names
                      dynamic_axes={
                          'speakers': {0: 'batch_size'},  # variable length axes
                          'phonemes': {0: 'batch_size', 1: 'seq_len'},
                          'src_lens': {0: 'batch_size'},
                          'output': {0: 'batch_size', 1: 'mel_len'},
                          'postnet_output': {0: 'batch_size', 1: 'mel_len'},
                          'p_predictions': {0: 'batch_size', 1: 'seq_len'},
                          'e_predictions': {0: 'batch_size', 1: 'seq_len'},
                          'log_d_predictions': {0: 'batch_size', 1: 'seq_len'},
                          'd_rounded': {0: 'batch_size', 1: 'seq_len'},
                          'src_masks': {0: 'batch_size', 1: 'seq_len'},
                          'mel_masks': {0: 'batch_size', 1: 'mel_len'},
                          'output_src_lens': {0: 'batch_size'},
                          'mel_lens': {0: 'batch_size'}
                      })
    print("done")
```

And actually, the output of the exported ONNX model is the same as pytorch inference output with model.eval() mode.

This is weird.

When I set training=torch.onnx.TrainingMode.EVAL, the exported ONNX model does not contain dropout layer, and when set training to TRAINING, the exported ONNX model does contain dropout layer, but generate same output each inference time, and these output are identical with the output of pytorch inference with model.eval() mode.

**Comment by jcwchen** [comment]
Hi @weixsong,
In your converted ONNX model, is there a seed (attribute) set in the Dropout node?

**Comment by weixsong** [comment]
> Hi @weixsong, In your converted ONNX model, is there a seed (attribute) set in the Dropout node?

No seed is set.

**Comment by jcwchen** [comment]
Thanks. Then what the training_mode and ratio arer for that Dropout node? If training_mode is true and ratio is not equal to 0, then something might be wrong in ONNXRuntime according to here: https://github.com/microsoft/onnxruntime/issues/5549#issuecomment-713758782. If training_mode is false, the output will always be the same. If possible, could you please provide your ONNX model? Than I can take a closer look. Thank you

**Comment by jcwchen** [comment]
Thank you for providing more details. I might know why -- could you please try to set different kinds of number for `onnxruntime.set_seed(x)` and see whether the output differs? It seems set_seed is designed for dropout in ONNXRuntime: https://github.com/microsoft/onnxruntime/pull/3490

**Comment by leng-yue** [comment]
I found a better solution :)

```python3
ort.InferenceSession("your_model", disabled_optimizers=["EliminateDropout"])
```

**Comment by weixsong** [comment]
> I found a better solution :)
> 
> ```python
> ort.InferenceSession("your_model", disabled_optimizers=["EliminateDropout"])
> ```

Yes, it works. Thanks very much.

**Comment by zhaozhiming37** [comment]
I have the same issue. But I can't solve it from the above infomation.When my last layer is `Dropout`, the output is as expected. But when I add `Flatten`, the output is weird. Part of my model is as follow.
![2022-07-28 17-52-10屏幕截图](https://user-images.githubusercontent.com/93253880/181479193-79adad6f-4f0d-4161-8049-e6aec3b3ca05.png)
![2022-07-28 17-52-52屏幕截图](https://user-images.githubusercontent.com/93253880/181479304-2472ff7e-f2b7-4e15-b073-45f75511362d.png)
I really don't know why, and I have tried to set `seed` for `ort`, and set `graph_optimization_level`, but not work at all.

**Comment by zhaozhiming37** [comment]
> I have the same issue. But I can't solve it from the above infomation.When my last layer is `Dropout`, the output is as expected. But when I add `Flatten`, the output is weird. Part of my model is as follow. ![2022-07-28 17-52-10屏幕截图](https://user-images.githubusercontent.com/93253880/181479193-79adad6f-4f0d-4161-8049-e6aec3b3ca05.png) ![2022-07-28 17-52-52屏幕截图](https://user-images.githubusercontent.com/93253880/181479304-2472ff7e-f2b7-4e15-b073-45f75511362d.png) I really don't know why, and I have tried to set `seed` for `ort`, and set `graph_optimization_level`, but not work at all.

@jcwchen

**Comment by ArchieLuxton** [comment]
I'm having the same issue with my model, except the above suggestions don't resolve the issue. My model is defined by:

`class MC_Dropout_Model(nn.Module):
    """
    A class to define the network implementing Dropout.
    Inference is performed later by a MC sampling method.
    """

    def __init__(self, input_dim, output_dim,
                 num_units, drop_prob, num_layers):
        super(MC_Dropout_Model, self).__init__()

        # input and output dimensions
        self.input_dim = input_dim
        self.output_dim = output_dim
        self.drop_prob = drop_prob
        self.num_layers = num_layers

        # n hidden layers
        layers = OrderedDict()

        # Input layer
        layers["0"] = nn.Linear(input_dim, num_units)

        for i in range(1, num_layers - 1):
            layers[str(i)] = nn.Linear(num_units, num_units)

        # Output layer
        layers[str(num_layers - 1)] = nn.Linear(num_units, output_dim)

        self.layers = nn.Sequential(layers)

    def forward(self, x):
        x = x.view(-1, self.input_dim)
        for layer in self.layers[:-1]:
            x = layer(x)
            x = torch.relu(x)
            x = F.dropout(x, p=self.drop_prob, training=True)
        x = self.layers[self.num_layers - 1](x)  # OUTPUT LAYER
        return x`

Thank you
